### PR TITLE
Adds a portion of liquidated collateral to CToken reserves

### PR DESCRIPTION
**Pull request by maxwolff**
Introduces `protocolSeizeShare`, the percentage of a liquidated account's seized collateral that will be added to reserves. This comes at the expense of the liquidator, who would normally earn the entire liquidated collateral amount.

Also adds `addReserves` to CEther, allowing this change to be implemented on CEther also.

### CODE REVIEW

**Review by jflatow — COMMENTED** [review]
Very nice - I think my only concern is if we are disincentivizing the most unhealthy liquidations from ever taking place (since the liquidator would lose money in the worst cases now as it gets whatever is leftover)

File: contracts/CToken.sol [hunk 1]
```diff
@@ -1058,52 +1070,48 @@ contract CToken is CTokenInterface, Exponential, TokenErrorReporter {
...
 
-        MathError mathErr;
-        uint borrowerTokensNew;
-        uint liquidatorTokensNew;
-        uint liquidatorSeizeTokens;
-        uint protocolSeizeTokens;
-        uint protocolSeizeAmount;
-        uint exchangeRateMantissa;
-        uint totalReservesNew;
+        SeizeInternalLocalVars memory vars;
```
Comment (maxwolff) [inline contracts/CToken.sol:1073]: stack too deep claims another ctoken fn ;/

File: contracts/CToken.sol [hunk 2]
```diff
@@ -1058,52 +1070,48 @@ contract CToken is CTokenInterface, Exponential, TokenErrorReporter {
...
-        require(mathErr == MathError.NO_ERROR, "exchange rate math error");
+        (vars.mathErr, vars.exchangeRateMantissa) = exchangeRateStoredInternal();
+        require(vars.mathErr == MathError.NO_ERROR, "exchange rate math error");
 
-        protocolSeizeAmount = mul_ScalarTruncate(Exp({mantissa: exchangeRateMantissa}), protocolSeizeTokens);
+        vars.protocolSeizeAmount = mul_ScalarTruncate(Exp({mantissa: vars.exchangeRateMantissa}), vars.protocolSeizeTokens);
 
-        totalReservesNew = add_(totalReserves, protocolSeizeAmount);
+        vars.totalReservesNew = add_(totalReserves, vars.protocolSeizeAmount);
+        vars.totalSupplyNew = sub_(totalSupply, vars.protocolSeizeTokens);
```
Comment (maxwolff) [inline contracts/CToken.sol:1094]: the seized tokens added to reserves are no longer part of the CToken supply

File: contracts/CToken.sol (hunk 1 above)
Comment (jflatow) [reply contracts/CToken.sol:1073]: :/

File: contracts/CToken.sol (hunk 2 above)
Comment (jflatow) [reply contracts/CToken.sol:1094]: 🎣

File: contracts/CEther.sol [hunk 3]
```diff
@@ -106,6 +106,14 @@ contract CEther is CToken {
         requireNoError(err, "liquidateBorrow failed");
     }
 
+    /**
+     * @notice The sender adds to reserves.
+     * @return uint 0=success, otherwise a failure (see ErrorReporter.sol for details)
+     */
+    function _addReserves() external payable returns (uint) {
```
[the remark this answers is not in the document]
Comment (maxwolff) [reply contracts/CEther.sol:113]: yes, at some point

**Pull request merged**